### PR TITLE
helm: Add serviceAccounts, rbac, and small fixes

### PR DESCRIPTION
**Pull request by jwitko**
### Description
* Add per-service serviceAccounts
* Add RBAC per service
* Fix historical commented config `druid_segmentCache_locations` to match PVC mount path. In it's current form it breaks the historical pod when uncommented due to lack of permissions to create directory.
* Update README.md with new options
* Add default annotation to historical service stateful set. Without any annotations the historical service breaks when `druid-kubernetes-extensions` is enabled.

### ATTENTION
This PR should be merged after #13746  since it contains a helm chart version bump incremental to that one.

#### Goal
The goal of this PR is to bring the apache/druid helm chart up to modern standards and fix some small issues. Enabling per-service serviceAccounts allows for finer grained RBAC which is in general a better security posture. Without this all services are forced to use the default serviceAccount which creates issues when needing to annotate service accounts for things like AWS iRSA as well as makes your only method of controlled permissions using a dedicated namespace.

#### Release note
Update suggested segment-cache path, Allow for per-service serviceAccounts in druid helm chart and finer-grained RBAC, and add a default annotation to historical statefulset.

This PR has:

- [X] been self-reviewed.
- [X] added documentation for new or modified features or behaviors.
- [X] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [X] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by abhishekagarwal87** [comment]
can you elaborate a bit more on "Enabling per-service serviceAccounts allows for finer grained RBAC"? Maybe take us through an example use-case that you want to implement on your own cluster.

**Comment by jwitko** [comment]
Merged upstream master into this branch

**Comment by jwitko** [comment]
> can you elaborate a bit more on "Enabling per-service serviceAccounts allows for finer grained RBAC"? Maybe take us through an example use-case that you want to implement on your own cluster.

Absolutely.
In its current state the helm chart does not allow for configuration of service accounts on the deployments/statefulsets.  This means your only choice is to use the `default` service account in whatever namespace you deploy into. Because of this, if you share the namespace with any other applications and they also use the default namespace you are now forced to share any roles/rolebindings that affect the `default` service account.  So if you want to isolate permissions this leaves your only option as putting druid in its own namespace so that nothing else shares the default service account.

Assuming you take the above approach and druid is now in its own namespace you have another issue.  Different druid services require different levels of permissions.  For example the `broker` does not require the ability to start kubernetes jobs but the `overlord` service does if you're using the `druid-kubernetes-overlord-extensions` extension.  With a single service account shared for all services you have no choice but to give all the services the ability to start jobs if you want to use the `druid-kubernetes-overlord-extensions` which goes against the principal of least privilege. You can extrapolate this even further by bringing the `druid-s3-extensions` into the conversation.  If you want to leverage [AWS iRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) for a better security posture and so that you don't have to store AWS credentials in your cluster and find a way to securely get them there you now have to give every single druid process the same IAM privileges in AWS if they are all using the same service account.

By merging this PR we get per-service service accounts which allow us to apply Kubernetes native RBAC and additional external privileges (like iRSA) on a per-service level so that we don't have to give unnecessary privileges to services that don't require them.

@abhishekagarwal87 does that make sense?  Let me know if you have any questions or disagree with any of the above.

**Comment by abhishekagarwal87** [comment]
Thank you. I see that by default, helm charts are now going to create service accounts per service. so can we keep the default off in top-level `values.yaml`

**Comment by jwitko** [comment]
> Thank you. I see that by default, helm charts are now going to create service accounts per service. so can we keep the default off in top-level `values.yaml`

Sorry, I'm not sure what you mean here?  Keep the default for what off?

File: helm/druid/templates/broker/deployment.yaml [hunk 1]
```diff
@@ -41,11 +41,15 @@ spec:
...
         component: {{ .Values.broker.name }}
-      {{- with .Values.broker.podAnnotations }}
       annotations:
-{{ toYaml . | indent 8 }}
+        druid.k8s.enablePatching: "true"
+      {{- with .Values.broker.podAnnotations }}
+        {{- toYaml . | nindent 8 }}
       {{- end }}
     spec:
+      {{- if .Values.broker.serviceAccount.create }}
```
Comment (dampcake) [inline helm/druid/templates/broker/deployment.yaml:50]: This blocks external creation of the serviceAccount. Most other charts allow you to bring your own serviceAccount which may also make reusing the same one for all druid service types easier.

I am wondering if we should default serviceAccount.name to unset and if create is true it uses what the default would be. Then if serviceAccount.name is set and serviceAccount.create is false we can still set the value here.

The charts I am used to using have a template like:

```
{{/*
Create the name of the forwarder service account to use
*/}}
{{- define "fluentd.forwarder.serviceAccountName" -}}
{{- if .Values.forwarder.serviceAccount.create -}}
    {{ default (printf "%s-forwarder" (include "common.names.fullname" .)) .Values.forwarder.serviceAccount.name }}
{{- else -}}
    {{ default "default" .Values.forwarder.serviceAccount.name }}
{{- end -}}
{{- end -}}
```

and then always set the serviceAccountName with the output of the template.

File: helm/druid/templates/broker/role.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{{- if .Values.rbac.create }}
```
Comment (dampcake) [inline helm/druid/templates/broker/role.yaml:1]: Should this also be checking `.Values.broker.enabled`? Feels like disabling the broker should also disable these resources from being created.

File: helm/druid/templates/broker/roleBinding.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+{{- if .Values.rbac.create }}
```
Comment (dampcake) [inline helm/druid/templates/broker/roleBinding.yaml:1]: Same question about checking enabled for the druid service, eg: `.Values.broker.enabled`

File: helm/druid/templates/broker/serviceAccount.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+{{- if .Values.broker.serviceAccount.create }}
```
Comment (dampcake) [inline helm/druid/templates/broker/serviceAccount.yaml:1]: Same question about checking enabled for the druid service, eg: `.Values.broker.enabled`

File: helm/druid/templates/broker/role.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+rules:
+  - apiGroups:
+      - ""
+    resources:
+      - pods
+      - configmaps
+    verbs:
+      - '*'
```
Comment (dampcake) [inline helm/druid/templates/broker/role.yaml:19]: Unclear on why the broker needs access to delete Pods, or really any of these. Are these permissions only needed by the Overlord?

Comment (jwitko) [reply helm/druid/templates/broker/role.yaml:19]: I found that when this role did not exist none of the pods could start.  They seem to all manage labels when using the kubernetes extension.  Unclear if they will remove when not needed the same as they add when needed.  I can test that and narrow down permission sets for each service.

File: helm/druid/templates/broker/deployment.yaml (hunk 1 above)
Comment (jwitko) [reply helm/druid/templates/broker/deployment.yaml:50]: I see what you're saying.  I have a bit of an opposite experience from you in terms of what other public widely used charts offer which is why I went this way but I'm not married to it.  I can create a helper definition which does the above sort of style.

Comment (jwitko) [reply helm/druid/templates/broker/deployment.yaml:50]: Fixed

File: helm/druid/templates/broker/role.yaml (hunk 2 above)
Comment (jwitko) [reply helm/druid/templates/broker/role.yaml]: Fixed

File: helm/druid/templates/broker/roleBinding.yaml (hunk 3 above)
Comment (jwitko) [reply helm/druid/templates/broker/roleBinding.yaml]: Fixed

File: helm/druid/templates/broker/serviceAccount.yaml (hunk 4 above)
Comment (jwitko) [reply helm/druid/templates/broker/serviceAccount.yaml]: Fixed

File: helm/druid/templates/broker/role.yaml (hunk 5 above)
Comment (jwitko) [reply helm/druid/templates/broker/role.yaml:19]: I pushed up the changes requested except for the RBAC changes since I haven't actually been able to say I've had a successful deployment with these chart changes yet.

Comment (jwitko) [reply helm/druid/templates/broker/role.yaml:19]: Turns out this is a bug with the kubernetes overlord extension and kubernetes 1.25+

File: helm/druid/README.md [hunk 6]
```diff
@@ -96,10 +96,16 @@ The following table lists the configurable parameters of the Druid chart and the
...
 | `gCloudStorage.secretName`               | secretName to be mounted as google cloud credentials    | `false`                                    |
+| `rbac.create`                            | Create roles and roleBindings for service Accounts      | `true`                                     |
 | `broker.enabled`                         | enable broker                                           | `true`                                     |
 | `broker.name`                            | broker component name                                   | `broker`                                   |
 | `broker.replicaCount`                    | broker node replicas (deployment)                       | `1`                                        |
 | `broker.port`                            | port of broker component                                | `8082`                                     |
+| `broker.serviceAccount.create`           | Create a service account for broker service             | `true`                                     |
+| `broker.serviceAccount.name`             | Service account name                                    | `true`                                     |
```
Comment (dampcake) [inline helm/druid/README.md:105]: The documentation looks off to me. Sorry I didn't catch this before. Looks like it says the default for everything is `true` but that's not actually the case.

File: helm/druid/templates/broker/role.yaml (hunk 5 above)
Comment (dampcake) [reply helm/druid/templates/broker/role.yaml:19]: Were you able to file an issue for it?

Comment (jwitko) [reply helm/druid/templates/broker/role.yaml:19]: Discovered this with the author of the kubernetes overlord extension @churromorales .  I believe https://github.com/apache/druid/pull/13759 already fixes the issues and relates to https://github.com/apache/druid/issues/13749 but maybe @churromorales can comment further if another issue is required I will happily make one.

File: helm/druid/README.md (hunk 6 above)
Comment (jwitko) [reply helm/druid/README.md]: No apologies, this is my fault.  Sorry I missed it during pre-commit review. Fixed the documentation and merged upstream/master again.

File: helm/druid/templates/broker/role.yaml (hunk 5 above)
Comment (jwitko) [reply helm/druid/templates/broker/role.yaml:19]: @dampcake putting aside the potential for a new issue to be created I would ask that we merge the RBAC in its current state since it matches the official documentation.  Once the bugs are fixed for K8s 1.25 I will test for the minimum viable role permissions and make a new PR to apply those via the helm chart as well as allow for concatenating or overriding the role permissions via the chart `.Values.rbac` map so people won't get blocked on changes required for rbac in the future.

If you're good with this let me know and I'll create the issue and it can be assigned to me.

**Review by dampcake — APPROVED** [review]
LGTM. Thanks for this!

File: helm/druid/values.yaml [hunk 7]
```diff
@@ -193,7 +233,7 @@ historical:
     druid_processing_numMergeBuffers: 2
     druid_processing_numThreads: 1
     # druid_monitoring_monitors: '["org.apache.druid.client.cache.CacheMonitor", "org.apache.druid.server.metrics.HistoricalMetricsMonitor", "org.apache.druid.server.metrics.QueryCountStatsMonitor"]'
-    # druid_segmentCache_locations: '[{"path":"/var/druid/segment-cache","maxSize":300000000000}]'
+    # druid_segmentCache_locations: '[{"path":"/opt/druid/var/druid/segment-cache","maxSize":300000000000}]'
```
Comment (churromorales) [inline helm/druid/values.yaml:236]: why the change here?

Comment (jwitko) [reply helm/druid/values.yaml:236]: That is the bullet point in the description 
```
Fix historical commented config druid_segmentCache_locations to match PVC mount path. In it's current form it breaks the historical pod when uncommented due to lack of permissions to create directory.
```
The container did not have permissions to create/write to `/var/druid/segment-cache`.  `/opt/druid/var/driuid/segment-cache` is inside of the defined PVC mounted volume path and so the application would have the ability to write there as well it will be captured by the persistent volume and persist through pod restarts.

Comment (churromorales) [reply helm/druid/values.yaml:236]: that makes sense, thank you for the fix.

**Review by churromorales — APPROVED** [review]

File: helm/druid/README.md [hunk 8]
```diff
@@ -199,6 +225,11 @@ The following table lists the configurable parameters of the Druid chart and the
 | `router.replicaCount`                    | router node replicas (deployment)                       | `1`                                        |
 | `router.port`                            | port of router component                                | `8888`                                     |
 | `router.serviceType`                     | service type for service                                | `ClusterIP`                                |
+| `router.serviceAccount.create`           | Create a service account for router service             | `true`                                     |
+| `router.serviceAccount.name`             | Service account name                                    | ``                                         |
```
Comment (abhishekagarwal87) [inline helm/druid/README.md:229]: what is the default service account name generated by helm? we should put that here instead of an empty string as default.

File: helm/druid/templates/historical/roleBinding.yaml [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    component: {{ .Values.historical.name }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: {{ template "druid.historical.fullname" . }}
+subjects:
+  - kind: ServiceAccount
+    name: {{ .Values.historical.serviceAccount.name }}
```
Comment (abhishekagarwal87) [inline helm/druid/templates/historical/roleBinding.yaml:18]: shouldn't this be `druid.historical.serviceAccountName`? The default value of this property is empty. is the chart generating the k8s configs correctly?

File: helm/druid/README.md (hunk 8 above)
Comment (jwitko) [reply helm/druid/README.md:229]: The default name is template derived so there is no way to put a value here that is guaranteed to be accurate.  It's a combination of the `druid.<service>.fullname` template which is based off the `druid.fullname` template and the service name. Most of which can be overridden.
```
{{- define "druid.fullname" -}}
{{- if .Values.fullnameOverride -}}
{{- .Values.fullnameOverride | trunc 63 | trimSuffix "-" -}}
{{- else -}}
{{- $name := default .Chart.Name .Values.nameOverride -}}
{{- if contains $name .Release.Name -}}
{{- .Release.Name | trunc 63 | trimSuffix "-" -}}
{{- else -}}
{{- printf "%s-%s" .Release.Name $name | trunc 63 | trimSuffix "-" -}}
{{- end -}}
{{- end -}}
{{- end -}}
```

```
{{- define "druid.broker.fullname" -}}
{{ template "druid.fullname" . }}-{{ .Values.broker.name }}
{{- end -}}
```

```
{{/*
Create the name of the broker service account
*/}}
{{- define "druid.broker.serviceAccountName" -}}
  {{- if .Values.broker.serviceAccount.create }}
    {{- default (include "druid.broker.fullname" .) .Values.broker.serviceAccount.name }}
  {{- else }}
    {{- default "default" .Values.broker.serviceAccount.name }}
  {{- end }}
{{- end }}
```

File: helm/druid/templates/historical/roleBinding.yaml (hunk 9 above)
Comment (jwitko) [reply helm/druid/templates/historical/roleBinding.yaml]: Thank you for catching this.  This was a left-over from before @dampcake recommended switching to the empty default value above in this PR and my testing had already had the `.Values.*.serviceAccount.name` values set from previous as well so I wasn't getting an issue here.

Fixed now.

File: helm/druid/README.md (hunk 8 above)
Comment (abhishekagarwal87) [reply helm/druid/README.md:229]: You can put `Derived from the name of service`. or something to that effect.

Comment (jwitko) [reply helm/druid/README.md]: Sure, fixed.

**Comment by abhishekagarwal87** [comment]
@jwitko - can you fix the static checks failures? LGTM otherwise.

**Comment by jwitko** [comment]
> @jwitko - can you fix the static checks failures? LGTM otherwise.

I accidentally rebased again forgetting that you guys prefer a merge instead.  trying to fix that now.

**Comment by jwitko** [comment]
@abhishekagarwal87 my apologies but I don't think I'm able to get the history back in a meaningful way to github.  If you want me to close this PR and just open a new one then please let me know.  Otherwise the PR looks to be in solid shape now with tests passing.

**Comment by jwitko** [comment]
@abhishekagarwal87 The static checks are failing for a spelling error that is outside the scope of this PR:
```
> spellcheck
> mdspell --en-us --ignore-numbers --report '../docs/**/*.md' || (./script/notify-spellcheck-issues && false)

    ../docs/tutorials/tutorial-jupyter-index.md
       41 | er both try to use port `8888,` so start Jupyter on a different port 

>> 1 spelling error found in 227 files
```
They seem to have passed on all things related to this chart once I added the license everywhere.

It seems to be complaining about [this line](https://github.com/apache/druid/blob/master/docs/tutorials/tutorial-jupyter-index.md?plain=1#L41) but I'm not really sure what its even upset about

**Review by abhishekagarwal87 — APPROVED** [review]

**Comment by abhishekagarwal87** [comment]
thanks @jwitko. Thats an unrelated failure.
